File: artifacts/api-server/src/lib/sdks/hexauth.js
/**
 * hexauth.js — Hex Auth Node.js SDK
 * Version: 1.0.0
 * Docs: https://hexauth.app/docs
 *
 * Requires Node.js 18+ (uses native fetch)
 * No external dependencies.
 *
 * npm: Copy this file into your project, or:
 *   const { HexAuth } = require("./hexauth");
 */

"use strict";

const os = require("os");
const crypto = require("crypto");

/**
 * Generate a hardware identifier for the current machine.
 * @returns {string} 32-char hex string
 */
function getHwid() {
  const parts = [
    os.hostname(),
    os.arch(),
    os.cpus()[0]?.model ?? "",
    String(os.totalmem()),
  ].join("|");
  return crypto.createHash("sha256").update(parts).digest("hex").slice(0, 32);
}

class HexAuthError extends Error {
  constructor(message) {
    super(message);
    this.name = "HexAuthError";
  }
}

class LoginResult {
  /**
   * @param {{ ok: boolean; message?: string; sessionToken?: string; user?: { username: string; plan: string; expiresAt?: string } }} data
   */
  constructor(data) {
    this.ok = Boolean(data.ok);
    this.message = data.message ?? "";
    this.sessionToken = data.sessionToken ?? null;
    this.user = data.user ?? null;
  }

  get username() { return this.user?.username ?? null; }
  get plan()     { return this.user?.plan ?? null; }

  toString() {
    return `LoginResult { ok: ${this.ok}, user: ${this.user?.username} }`;
  }
}

class Variable {
  /**
   * @param {{ name: string; value: string }} data
   */
  constructor(data) {
    this.name = data.name ?? "";
    this.value = data.value ?? "";
  }
}

class HexAuth {
  /**
   * @param {{ appId: string; appSecret: string; version?: string; endpoint?: string; timeout?: number }} options
   *
   * @example
   * const api = new HexAuth({
   *   appId: "YOUR_APP_ID",
   *   appSecret: "YOUR_SECRET",
   *   version: "1.0",
   * });
   */
  constructor({ appId, appSecret, version = "1.0", endpoint = "https://hexauth.app/api", timeout = 10000 } = {}) {
    if (!appId || !appSecret) throw new HexAuthError("appId and appSecret are required");
    this._appId = appId;
    this._appSecret = appSecret;
    this._version = version;
    this._endpoint = endpoint.replace(/\/+$/, "");
    this._timeout = timeout;
    this._hwid = getHwid();
    this._sessionToken = null;
  }

  async _fetch(method, path, body, token) {
    const url = `${this._endpoint}${path}`;
    const headers = {
      "Content-Type": "application/json",
      "User-Agent": `HexAuth-Node/1.0 v${this._version}`,
    };
    if (token) headers["Authorization"] = `Bearer ${token}`;

    const ctrl = new AbortController();
    const timer = setTimeout(() => ctrl.abort(), this._timeout);
    try {
      const resp = await fetch(url, {
        method,
        headers,
        body: body ? JSON.stringify(body) : undefined,
        signal: ctrl.signal,
      });
      clearTimeout(timer);
      return await resp.json();
    } catch (err) {
      clearTimeout(timer);
      return { ok: false, message: err.name === "AbortError" ? "Request timed out" : `Network error: ${err.message}` };
    }
  }

  /**
   * Authenticate a user. Binds HWID on first login.
   * @param {{ username: string; password: string }} opts
   * @returns {Promise<LoginResult>}
   */
  async login({ username, password }) {
    const data = await this._fetch("POST", "/sdk/login", {
      appId: this._appId,
      username,
      password,
      hwid: this._hwid,
      version: this._version,
    });
    const result = new LoginResult(data);
    if (result.ok && result.sessionToken) this._sessionToken = result.sessionToken;
    return result;
  }

  /**
   * Validate an existing session. Uses last login token if not provided.
   * @param {string} [sessionToken]
   * @returns {Promise<LoginResult>}
   */
  async validate(sessionToken) {
    const token = sessionToken ?? this._sessionToken;
    if (!token) return new LoginResult({ ok: false, message: "No session token" });
    const data = await this._fetch("POST", "/sdk/validate", { sessionToken: token });
    return new LoginResult(data);
  }

  /**
   * Fetch a remote variable by name. Requires active session.
   * @param {string} name
   * @param {string} [sessionToken]
   * @returns {Promise<Variable>}
   */
  async getVariable(name, sessionToken) {
    const token = sessionToken ?? this._sessionToken;
    if (!token) throw new HexAuthError("Not logged in. Call login() first.");
    const data = await this._fetch("GET", `/sdk/variable/${encodeURIComponent(name)}`, null, token);
    return new Variable(data);
  }

  /** @returns {string} Hardware ID of the current machine */
  static getHwid() { return getHwid(); }

  /** @returns {string | null} Current session token */
  get sessionToken() { return this._sessionToken; }
}

module.exports = { HexAuth, LoginResult, Variable, HexAuthError, getHwid };

// ── Quick-start example ──────────────────────────────────────────────────────
// const { HexAuth } = require("./hexauth");
//
// const api = new HexAuth({
//   appId: "YOUR_APP_ID",
//   appSecret: "YOUR_SECRET",
//   version: "1.0",
// });
//
// const result = await api.login({ username: "developer123", password: "secure_password" });
// if (result.ok) {
//   console.log(`Welcome ${result.user.username}! Plan: ${result.user.plan}`);
//   const v = await api.getVariable("download_url");
//   console.log(`Download: ${v.value}`);
// } else {
//   console.log(`Auth failed: ${result.message}`);
// }

File: artifacts/api-server/src/lib/sdks/hexauth.py
"""
hexauth.py — Hex Auth Python SDK
Version: 1.0.0
Docs: https://hexauth.app/docs
"""

import hashlib
import platform
import socket
import uuid
import urllib.request
import urllib.error
import json
import sys
import subprocess
from typing import Optional


def _get_hwid() -> str:
    """Generate a unique hardware identifier for this machine."""
    try:
        parts = [
            platform.node(),
            platform.machine(),
            platform.processor(),
            str(uuid.getnode()),
        ]
        # Try to get disk serial on Windows
        if sys.platform == "win32":
            try:
                result = subprocess.check_output(
                    "wmic diskdrive get SerialNumber",
                    shell=True, stderr=subprocess.DEVNULL
                ).decode().strip().split()
                if len(result) > 1:
                    parts.append(result[-1])
            except Exception:
                pass
        raw = "|".join(p for p in parts if p)
        return hashlib.sha256(raw.encode()).hexdigest()[:32]
    except Exception:
        return hashlib.sha256(platform.node().encode()).hexdigest()[:32]


class User:
    def __init__(self, data: dict):
        self.username: str = data.get("username", "")
        self.plan: str = data.get("plan", "free")
        self.expires_at: Optional[str] = data.get("expiresAt")

    def __repr__(self):
        return f"<User username={self.username!r} plan={self.plan!r}>"


class Variable:
    def __init__(self, data: dict):
        self.name: str = data.get("name", "")
        self.value: str = data.get("value", "")

    def __repr__(self):
        return f"<Variable {self.name!r}={self.value!r}>"


class LoginResult:
    def __init__(self, data: dict):
        self.ok: bool = data.get("ok", False)
        self.message: str = data.get("message", "")
        self.session_token: Optional[str] = data.get("sessionToken")
        self.user: Optional[User] = User(data["user"]) if data.get("user") else None

    def __bool__(self):
        return self.ok

    def __repr__(self):
        return f"<LoginResult ok={self.ok} user={self.user}>"


class HexAuthError(Exception):
    pass


class HexAuth:
    """
    Hex Auth SDK client.

    Usage:
        from hexauth import HexAuth

        api = HexAuth(
            app_id="YOUR_APP_ID",
            app_secret="YOUR_SECRET",
            version="1.0"
        )

        result = api.login("username", "password")
        if result.ok:
            print(f"Welcome {result.user.username}!")
    """

    def __init__(
        self,
        app_id: str,
        app_secret: str,
        version: str = "1.0",
        endpoint: str = "https://hexauth.app/api",
        timeout: int = 10,
    ):
        self.app_id = app_id
        self.app_secret = app_secret
        self.version = version
        self.endpoint = endpoint.rstrip("/")
        self.timeout = timeout
        self._session_token: Optional[str] = None
        self._hwid: str = _get_hwid()

    def _request(self, method: str, path: str, body: Optional[dict] = None, token: Optional[str] = None) -> dict:
        url = f"{self.endpoint}{path}"
        data = json.dumps(body).encode() if body else None
        headers = {
            "Content-Type": "application/json",
            "User-Agent": f"HexAuth-Python/1.0 v{self.version}",
        }
        if token:
            headers["Authorization"] = f"Bearer {token}"

        req = urllib.request.Request(url, data=data, headers=headers, method=method)
        try:
            with urllib.request.urlopen(req, timeout=self.timeout) as resp:
                return json.loads(resp.read().decode())
        except urllib.error.HTTPError as e:
            try:
                err = json.loads(e.read().decode())
                return {"ok": False, "message": err.get("message", str(e))}
            except Exception:
                return {"ok": False, "message": str(e)}
        except urllib.error.URLError as e:
            return {"ok": False, "message": f"Network error: {e.reason}"}

    def login(self, username: str, password: str) -> LoginResult:
        """
        Authenticate a user. Binds HWID on first login.

        Returns:
            LoginResult with .ok, .user, .session_token, .message
        """
        payload = {
            "appId": self.app_id,
            "username": username,
            "password": password,
            "hwid": self._hwid,
            "version": self.version,
        }
        data = self._request("POST", "/sdk/login", payload)
        result = LoginResult(data)
        if result.ok and result.session_token:
            self._session_token = result.session_token
        return result

    def validate(self, session_token: Optional[str] = None) -> LoginResult:
        """
        Validate an existing session token.
        Uses the token from the last login() call if not provided.
        """
        token = session_token or self._session_token
        if not token:
            return LoginResult({"ok": False, "message": "No session token"})
        data = self._request("POST", "/sdk/validate", {"sessionToken": token})
        return LoginResult(data)

    def get_variable(self, name: str, session_token: Optional[str] = None) -> Variable:
        """
        Fetch a remote variable by name.
        Requires an active session.
        """
        token = session_token or self._session_token
        if not token:
            raise HexAuthError("Not logged in. Call login() first.")
        data = self._request("GET", f"/sdk/variable/{name}", token=token)
        return Variable(data)

    @staticmethod
    def get_hwid() -> str:
        """Return the hardware ID of the current machine."""
        return _get_hwid()


# ── Quick-start example ──────────────────────────────────────────────────────
if __name__ == "__main__":
    import getpass

    api = HexAuth(
        app_id=input("App ID: "),
        app_secret=input("App Secret: "),
        version="1.0",
    )

    username = input("Username: ")
    password = getpass.getpass("Password: ")

    result = api.login(username, password)
    if result.ok:
        print(f"\n✓ Welcome {result.user.username}! Plan: {result.user.plan}")
    else:
        print(f"\n✗ Login failed: {result.message}")
        sys.exit(1)
